src/utils/communication_buffer.hpp

```diff
@@ -0,0 +1,418 @@
+  CommBuffer(const CommBuffer<U> &in);
+
+  template <class U>
+  CommBuffer &operator=(const CommBuffer<U> &in);
+
+  operator T &() { return buf_; }
+  operator const T &() const { return buf_; }
+
+  T &buffer() { return buf_; }
+  const T &buffer() const { return buf_; }
+
+  void Allocate() {
+    if (!active_) {
+      buf_ = get_resource_();
+      active_ = true;
+    }
+  }
+
+  void Free() {
+    buf_ = T();
+    active_ = false;
+  }
+
+  bool IsActive() const { return active_; }
+
+  BufferState GetState() { return *state_; }
+
+  void Send() noexcept;
+  void SendNull() noexcept;
+
+  bool IsAvailableForWrite();
+
+  void TryStartReceive() noexcept;
+  bool TryReceive() noexcept;
+
+  void Stale();
+};
+
+// Method definitions below
+
+template <class T>
+CommBuffer<T>::CommBuffer(int tag, int send_rank, int recv_rank, comm_t comm,
+                          std::function<T()> get_resource, bool do_sparse_allocation)
+    : state_(std::make_shared<BufferState>(BufferState::stale)),
+      comm_type_(std::make_shared<BuffCommType>(BuffCommType::both)),
+      started_irecv_(std::make_shared<bool>(false)),
+      nrecv_tries_(std::make_shared<int>(0)),
+#ifdef MPI_PARALLEL
+      my_request_(std::make_shared<MPI_Request>(MPI_REQUEST_NULL)),
+#endif
+      tag_(tag), send_rank_(send_rank), recv_rank_(recv_rank), comm_(comm),
+      get_resource_(get_resource), buf_() {
+// Set up persistent communication
+#ifdef MPI_PARALLEL
+  PARTHENON_MPI_CHECK(MPI_Comm_rank(comm_, &my_rank));
+#else
+  my_rank = 0;
+#endif
+  if (send_rank == recv_rank) {
+    assert(my_rank == send_rank);
+    *comm_type_ = BuffCommType::both;
+  } else if (my_rank == send_rank) {
+    *comm_type_ = BuffCommType::sender;
+  } else if (my_rank == recv_rank) {
+    *comm_type_ = BuffCommType::receiver;
+    if (do_sparse_allocation) *comm_type_ = BuffCommType::sparse_receiver;
+  } else {
+    // This is an error
+    std::cout << "CommBuffer initialization error" << std::endl;
+  }
+}
+
+template <class T>
+template <class U>
+CommBuffer<T>::CommBuffer(const CommBuffer<U> &in)
+    : buf_(in.buf_), state_(in.state_), comm_type_(in.comm_type_),
+      started_irecv_(in.started_irecv_), nrecv_tries_(in.nrecv_tries_),
+      my_request_(in.my_request_), tag_(in.tag_), send_rank_(in.send_rank_),
+      recv_rank_(in.recv_rank_), comm_(in.comm_), active_(in.active_) {
+#ifdef MPI_PARALLEL
+  PARTHENON_MPI_CHECK(MPI_Comm_rank(comm_, &my_rank));
+#else
+  my_rank = 0;
+#endif
+}
+
+template <class T>
+CommBuffer<T>::~CommBuffer() {
+#ifdef MPI_PARALLEL
+  if (my_request_.use_count() == 1) { // This is the last shallow copy of this buffer
+    int flag;
+    MPI_Status status;
+    PARTHENON_MPI_CHECK(MPI_Test(my_request_.get(), &flag, &status));
+    if (!flag) {
+      if (*comm_type_ == BuffCommType::sender) {
+        PARTHENON_MPI_CHECK(MPI_Wait(my_request_.get(), MPI_STATUS_IGNORE));
+      } else {
+        PARTHENON_MPI_CHECK(MPI_Cancel(my_request_.get()));
```

Comment:
Is this line the equivalent of the previous `ClearBoundary` calls?

Reply:
I am not sure, but I think `ClearBoundary` is more like calling `Stale` on a buffer. This is just making sure there are no requests flying around for buffers that have been destroyed (as would happen when remeshing occurs).

Reply:
Maybe add this as a comment above the line?

Reply:
comment added